src/Servers/Kestrel/Core/src/Internal/Http/HttpProtocol.cs

```diff
@@ -275,6 +277,23 @@ public CancellationToken RequestAborted
 
         protected HttpResponseHeaders HttpResponseHeaders { get; } = new HttpResponseHeaders();
 
+        HttpContext IHttpContextContainer.HttpContext
+        {
+            get
+            {
+                if (_httpContext is null)
```

Comment:
Hmm this change might be a bit dangerous if this is accessed during the request, it'll re-initialize?

Reply:
Do you mean if it's accessed outside of the request flow?

Reply:
Inside. If you get the HttpContext and cast the FeatureCollection property to IHttpContextContainer then it’ll call Initialize again

Reply:
Create it upfront in the variable
```
private ReusableHttpContext _httpContext = new ReusableHttpContext(this);
```
Add an `IsInitialized` flag in `ReusableHttpContext` and use that in the property?

Reply:
I'm not sure I understand what you're getting at. You mean something like the following?

```C#
app.Run(async context =>
{
    var context2 = ((IHttpContextContainer)context.Features).HttpContext;
    await context.Response.WriteAsync(object.ReferenceEquals(context, context2).ToString());
});
```

In that case the response would be "True". I'm not sure why you'd think this wouldn't be the case.

I'm also not sure why we're concerned about what happens when the IFeatureCollection to IHttpContextContainer. I don't know why we'd have any burden whatsoever to support that.

Reply:
It's super edge case. I'm going to merge this anyways, I just wanted to bring it up.